### DISCUSSION TITLE
Other Language Support

### DISCUSSION THREAD

**Discussion by mLienhard** [aredden/electron-G14Control#41] [Ideas]
Hi together,
I'd like to start the discussion how the support for other (Windows Installation) Languages could look like.

There are some Lookup strings in the code which differes from Windows Language to Language.

My first Idea was to outsource alle these Language dependend strings to some kind of INI file (or XML File) and provide some load menu to load these from the GUI.

in this case it would be quite easy to add new languages and maintain them.

What are your thoughts for such a feature?

**Comment by aredden** [comment]
That actually sounds pretty reasonable. I could add a json / yml file with the values for the thermal / battery counters & powercfg related commands & then you could change them.

**Comment by mLienhard** [comment]
Sounds very good!
I'd like to keep one file for each (needed) langauge and select the used file by the configuration (json) file. Perhaps we could even find a possibility to auto detect the windows installation language. Then no manual switching/configuration would be needed in future... (of course once the corresponding translation file is created and filled)

**Reply by aredden** [reply]
Yeah detecting language probably wouldn't be that hard. I'm also currently (like my other open window is visual studio) working on another approach at least for powercfg stuff. I can set the boost and get boost levels via the pwrprof driver & I'll try to add that asap. I am still unsure about the performance counters though.

**Comment by mLienhard** [comment]
Hey,
I just saw you uploaded a relase with Beta Multilanguage support.
I tried to understand how this should happened but i failed!

I installed the latest version but it does not seem to work ( at least my temperature is still reported to be 0 (with german windows).
Is there something I've to add somewhere?
Or is there any possibility i can help you out with testing, logging, what ever?

**Reply by mLienhard** [reply]
this is how my startup output looks like with the latest version

[Main - INFO] - Running in development
[ConfigLoader - INFO] - Persistent configuration exists.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
[ConfigLoader - INFO] - Number of validation Errors: 0
[ConfigLoader - INFO] - Using appdata persistent config.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
(node:9104) ExtensionLoadWarning: Warnings loading extension at C:\Users\Markus\AppData\Roaming\G14ControlV2-Main\extensions\fmkadmapgofadopljbjfkapdkoienihi: Unrecognized manifest key 'browser_action'. Unrecognized manifest key 'minimum_chrome_version'. Unrecognized manifest key 'update_url'. Cannot load extension with file or directory name _metadata. Filenames starting with "_" are reserved for use by the system.
Added Extension:  React Developer Tools
[HIDControl - INFO] - Found G14 unique key
[Main - INFO] - Updating state of rogKey hid device.
[Main - INFO] - ROG key HID built and listening.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
[Main - INFO] - Renderer process built and updater being initialized...
[AppUpdater - INFO] - Updater Initialized.
[AppUpdater - INFO] - Checking for update.
[BoostVisibility - INFO] - RegistryChecker result: True

[StartupChecks - INFO] - CheckBoostVisibilityResult true
[IPCEmitters - INFO] - Perfmon loop started...
[WindowsPlanListeners - ERROR] - Parsed single line boost regex values were not found.
{"acboost":null,"dcboost":null}
[HardwareControl - INFO] - Which Charger Result: 0
[HardwareControl - INFO] - Which Charger Result: 0

**Reply by aredden** [reply]
Are you using the code or the .exe, because it looks like you are running the code, which is a bit confusing (you shouldn't see the extensions error if it's running in production). If you are running the .exe that I release from [here](https://github.com/aredden/electron-g14control/releases), then you need to do what I described in the release notes: switch the multilanguage support switch from the settings page and then restart the application. It's not guaranteed to work, and will only work for viewing temperature and (possibly but probably not) battery drain.

**Reply by aredden** [reply]
I just realized that you are running in development on purpose.. and therefore I feel dumb. So yeah all you would need to do is switch the multilang switch on the settings page and then restart the app. It should then use an ID to grab the name of the typeperf process that need to be listened to. The code I changed is here: https://github.com/aredden/electron-G14Control/blob/69f88c8ee0f9bb9aaab82cb0ce9629159e4c9e6c/electron/src/IPCEvents/TempPerf/Perfmon.ts#L54-L114

**Comment by mLienhard** [comment]
Hi,
i was not sure whether i really restarted the app in development mode with enabled multilanguage support.
Now im sure - unfortunately it is still not working.

From my point of view it looks like the ap does not reach the code which trys to load the perfmon at all:
The windows plans can also not be parsed from my annoying german windows...


[Main - INFO] - Running in development
[ConfigLoader - INFO] - Persistent configuration exists.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
[ConfigLoader - INFO] - Number of validation Errors: 0
[ConfigLoader - INFO] - Using appdata persistent config.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
(node:16988) ExtensionLoadWarning: Warnings loading extension at C:\Users\Markus\AppData\Roaming\G14ControlV2-Main\extensions\fmkadmapgofadopljbjfkapdkoienihi: Unrecognized manifest key 'browser_action'. Unrecognized manifest key 'minimum_chrome_version'. Unrecognized manifest key 'update_url'. Cannot load extension with file or directory name _metadata. Filenames starting with "_" are reserved for use by the system.
Added Extension:  React Developer Tools
[HIDControl - INFO] - Found G14 unique key
[Main - INFO] - Updating state of rogKey hid device.
[Main - INFO] - ROG key HID built and listening.
[ConfigLoader - INFO] - Loading config from:
[C:\Users\Markus\AppData\Roaming\G14ControlV2\G14ControlV2.json]
[Main - INFO] - Renderer process built and updater being initialized...
[AppUpdater - INFO] - Updater Initialized.
[AppUpdater - INFO] - Checking for update.
[BoostVisibility - INFO] - RegistryChecker result: True

[StartupChecks - INFO] - CheckBoostVisibilityResult true
[IPCEmitters - INFO] - Perfmon loop started...
[WindowsPlanListeners - ERROR] - Parsed single line boost regex values were not found.
{"acboost":null,"dcboost":null}
[HardwareControl - INFO] - Which Charger Result: 65537
[HardwareControl - INFO] - Which Charger Result: 65537
[HardwareControl - INFO] - Which Charger Result: 65537
[WindowsPlanListeners - ERROR] - Parsed single line boost regex values were not found.
{"acboost":null,"dcboost":null}
[HardwareControl - INFO] - Which Charger Result: 65537

**Reply by aredden** [reply]
Yeah the windows boost and graphics values still won't be able to be parsed.. but I have a fix for that -- I just have to stop being lazy and hook up the powercfg alternative c# cli mini-app that I already made. So getting WindowsPlanListener errors should be expected. The only thing that should change would be that you would be able to see the temperature in the top left box.
![image](https://user-images.githubusercontent.com/14303104/106359854-31f26700-62e3-11eb-815c-caf6926e4908.png)
The switch is here...
![image](https://user-images.githubusercontent.com/14303104/106359865-45053700-62e3-11eb-8cd3-c26d62a9f3ff.png)

**Comment by mLienhard** [comment]
This is what i alreasy tried but i thing it did not had the effect you expected (and i hoped)
(and yes: i already restartet the app)

![EnabledMultilanguage](https://user-images.githubusercontent.com/68334156/106366776-8d057780-633e-11eb-9629-ec21ecd91a00.PNG)

**Reply by mLienhard** [reply]
by the way, i would not call you lazy...
I tested very shortly the pwrprof-simple-cli and it looks like it is working as expected (even with a german windows... ;-) )

The reporting of the temperature is a small feature - I don't know why but I feel somehow uncomfortable when the tool i use to control the fan curves is not able to display the current temperature (although I know that the tool which really controls the fan curves is a different one and is working independend from the GUI we are talking)

**Reply by aredden** [reply]
That is strange. I guess it could be that the ID of the plans is different on your computer.. basically what I did was I used two powershell commands, (which you can add to your powershell by running `notepad $Profile` and then copying and pasting these there:

```ps

Function Get-PerformanceCounterLocalName
{
  param
  (
    [UInt32]
    $ID,
 
    $ComputerName = $env:COMPUTERNAME
  )
 
  $ComputerName = $env:COMPUTERNAME
  $code = '[DllImport("pdh.dll", SetLastError=true, CharSet=CharSet.Unicode)] public static extern UInt32 PdhLookupPerfNameByIndex(string szMachineName, uint dwNameIndex, System.Text.StringBuilder szNameBuffer, ref uint pcchNameBufferSize);'
 
  $Buffer = New-Object System.Text.StringBuilder(1024)
  [UInt32]$BufferSize = $Buffer.Capacity
 
  $t = Add-Type -MemberDefinition $code -PassThru -Name PerfCounter -Namespace Utility
  $rv = $t::PdhLookupPerfNameByIndex($ComputerName, $id, $Buffer, [Ref]$BufferSize)
 
  if ($rv -eq 0)
  {
    $Buffer.ToString().Substring(0, $BufferSize-1)
  }
  else
  {
    Throw 'Get-PerformanceCounterLocalName : Unable to retrieve localized name. Check computer name and performance counter ID.'
  }
}
 
function Get-PerformanceCounterID
{
    param
    (
        [Parameter(Mandatory=$true)]
        $Name
    )
 
    if ($script:perfHash -eq $null)
    {
        Write-Progress -Activity 'Retrieving PerfIDs' -Status 'Working'
 
        $key = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\CurrentLanguage'
        $counters = (Get-ItemProperty -Path $key -Name Counter).Counter
        $script:perfHash = @{}
        $all = $counters.Count
 
        for($i = 0; $i -lt $all; $i+=2)
        {
           Write-Progress -Activity 'Retrieving PerfIDs' -Status 'Working' -PercentComplete ($i*100/$all)
           $script:perfHash.$($counters[$i+1]) = $counters[$i]
        }
    }
 
    $script:perfHash.$Name
}
```

Then trying to find the ID for the alternative to the english version of these two counters:
'\Thermal Zone Information(*)\High Precision Temperature'
'\Power Meter(_Total)\Power'

**Reply by mLienhard** [reply]
Hi,
i tried your scripts but i did not get any result. I don't know whether this is part of the problem or not?!
the german names should be 
'\Thermozoneninformationen(*)\Hochpräzise Temperatur'
'\Energiemessung(_Total)\Leistung'

i tried to find the numbers manually in the registry and would think the Id should be:
High Precission Temperature = 3802
Power Meter \ Power = 6482

![PowerShell_PerformanceCounter](https://user-images.githubusercontent.com/68334156/106666297-b2e18500-65a7-11eb-9e96-07fa679fd762.PNG)

I thought the idea of these IDs is that they are language independend?! I'm confused...

**Reply by aredden** [reply]
It's language independent because (I assumed) that all windows machines would have the same ID's therefore I could find the language dependent name regardless of the language and then use that as the counter that I would grab temp & battery drain from.
